File: 0132. Palindrome Partitioning II/Solution0132.cpp
#include<iostream>
#include<vector>
#include<string>

using namespace std;

class Solution0132 {
public:
    int minCut(string s) {
        vector<int> dp(s.length(), 0);
        for (int i = 1; i < s.length(); i++) {
            dp[i] = i;
        }
        for (int mid = 0; mid < s.length(); mid++) {
            helper(s, mid-1, mid, dp);            
            helper(s, mid, mid, dp);
        }
        return dp[s.length() - 1];
    }
    
    void helper(string s, int left, int right, vector<int>& dp) {
        while (left >= 0 && right < s.length() && s[left] == s[right]) {
            int rangeMinCut = left == 0 ? 0 : dp[left-1] + 1;
            dp[right] = min(dp[right], rangeMinCut);
            left--; right++;
        }
    }
};

File: 0132. Palindrome Partitioning II/Solution0132.py
class Solution0132:
    def minCut(self, s: str) -> int:
        dp = [i for i in range(len(s))]
        for mid in range(len(s)):
            self.helper(s, mid-1, mid, dp)            
            self.helper(s, mid, mid, dp)
        return dp[len(s)-1]
    
    def helper(self, s:str, left: int, right: int, dp: List[int]) -> None:
        while left >= 0 and right < len(s) and s[left] == s[right]:
            range_min_cut = 0 if left == 0 else (dp[left-1] + 1)
            dp[right] = min(dp[right], range_min_cut)
            left-=1
            right+=1